### PR TITLE
meta: architectures

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `make lint`?
- [X] Have you successfully run `pytest tests/unit`?

-----
## Overview

Add support for architectures.

- Snapcraft's lifecycle will execute for each `build_on`->`build_for` defined in `snapcraft.yaml`.  
- Only snaps where `build_on` equals the host architecture will be built.
- When using `lxd` or `multipass`, a separate instance will be used for each `build_on`->`build_for` pair.

7th and final part of supporting architectures for `core22`. 

## Details
- Add support for `build-on` and `build-for` for `core22` snaps.
- Add environmental variable `SNAPCRAFT_BUILD_FOR` to limit which snaps are built

Note: For `core22` snaps, `run-on` has been replaced with `build-for`.

### Examples
These example `snapcraft.yaml` snippets show how to use architectures.

#### Single architecture
Building on `amd64` will produce 1 snap for `amd64`.
```
architectures:
  - build-on: amd64
    build-for: amd64
```

#### Single architecture, shorthand
This is the same as the previous example, but in a shorthand format.
```
architectures:
  - amd64
```

#### Multiple architectures
- Building on `amd64` will produce 1 snap for `amd64`.
- Building on `arm64` will produce 1 snap for `arm64`.
```
architectures:
  - build-on: amd64
    build-for: amd64
  - build-on: arm64
    build-for: arm64
```

#### Multiple architectures, shorthand
This is the same as the previous example, but in a shorthand format.
```
architectures:
  - amd64
  - arm64
```

#### Multiple architecture, cross-compiling
- Building on `amd64` will produce 2 snaps, 1 snap for `amd64` and 1 snap for `arm64`.
- Building on `arm64` will produce 1 snap for `arm64`.
```
architectures:
  - build-on: amd64
    build-for: amd64
  - build-on: [amd64, arm64]
    build-for: arm64
```

#### Architecture independent
For snaps that can run on any architecture (e.g. python or shell scripts), use `build-for: all`.
```
architectures:
  - build-on: amd64
    build-for: all
```

#### Limiting builds with environmental variable
The environmental variable `SNAPCRAFT_BUILD_FOR` can be used so that only 1 snap package is produced.
- Running `snapcraft` on `amd64` will produce 2 snaps, 1 snap for `amd64` and 1 snap for `arm64`.
- Running `SNAPCRAFT_BUILD_FOR=arm64 snapcraft` will produce 1 snap for `arm64`.
```
architectures:
  - build-on: amd64
    build-for: amd64
  - build-on: amd64
    build-for: arm64
```

(CRAFT-1173)

### CODE REVIEW

**Comment by mr-cal** [comment]
I introduced a regression this week where snaps are being built more than once.
I'm moving this back to draft so I can fix it.

File: snapcraft/parts/lifecycle.py [hunk 1]
```diff
@@ -317,7 +344,12 @@ def _run_in_provider(
     provider = providers.get_provider(provider_name)
     provider.ensure_provider_is_available()
```
Comment (mr-cal) [inline snapcraft/parts/lifecycle.py:346]: `def _run_in_provider()` has no unit tests :(

After trying to write the unit tests this afternoon, I understand why none exist.  Mocking the `Executor` objects is not trivial.  Charmcraft has a complex solution and I'm not good enough with pytest to get it working for snapcraft in an afternoon.

File: snapcraft/parts/lifecycle.py [hunk 2]
```diff
@@ -161,22 +186,30 @@ def run(command_name: str, parsed_args: "argparse.Namespace") -> None:
...
-
-    project = Project.unmarshal(yaml_data)
 
-    try:
-        _run_command(
-            command_name,
-            project=project,
-            parse_info=parse_info,
+    for build_on, build_for in build_plan:
+        emit.progress(f"running with build_on: {build_on} and build_for: {build_for}")
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/parts/lifecycle.py]: I took all 3 of your suggestions here

File: snapcraft/parts/lifecycle.py [hunk 3]
```diff
@@ -485,3 +522,50 @@ def _expand_environment(
...
+    build_plan: List[Tuple[str, str]] = []
+
+    # `isinstance()` calls are for mypy type checking and should not change logic
+    for arch in [arch for arch in archs if isinstance(arch, Architecture)]:
+        for build_on in arch.build_on:
+            if build_on in host_arch and isinstance(arch.build_for, list):
+                build_plan.append((host_arch, arch.build_for[0]))
+            else:
+                emit.progress(
+                    f"skipping build-on: {build_on} build-for: {arch.build_for}"
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/parts/lifecycle.py]: I changed this to a verbose message too.

File: snapcraft/parts/lifecycle.py [hunk 4]
```diff
@@ -366,7 +401,12 @@ def _run_in_provider(
     provider = providers.get_provider(provider_name)
     provider.ensure_provider_is_available()
 
-    cmd = ["snapcraft", command_name]
+    # pass `build-for` of the current build plan to the instance
+    cmd = [
+        f"SNAPCRAFT_BUILD_FOR={project.get_build_for()}",
```
[the remark this answers is not in the document]
Comment (mr-cal) [reply snapcraft/parts/lifecycle.py]: Good advice.  I wasn't aware of the paradigm used for `SNAPCRAFT_BUILD_INFO` and `SNAPCRAFT_IMAGE_INFO`.

I added a new argument, `--build-to <architecture>`, which defaults to `SNAPCRAFT_BUILD_FOR`.

File: snapcraft/parts/lifecycle.py [hunk 5]
```diff
@@ -23,17 +23,26 @@
...
+from snapcraft.projects import (
+    Architecture,
+    ArchitectureProject,
+    GrammarAwareProject,
+    Project,
+)
 from snapcraft.providers import capture_logs_from_instance
-from snapcraft.utils import get_host_architecture, process_version
+from snapcraft.utils import (
+    ARCH_TRANSLATIONS_DEB_TO_PLATFORM,
```
Comment (cmatsuoka) [inline snapcraft/parts/lifecycle.py:42]: This is still importing the mapping directly, any special reason for not using the function here?

Comment (mr-cal) [reply snapcraft/parts/lifecycle.py]: Whoops!  I made the dictionaries private, added the function, and a note about 2 dictionaries.

**Review by cmatsuoka — APPROVED** [review]
Thanks Callahan!

**Pull request merged**